dojo_plugin/api/v1/dojo.py

```diff
@@ -0,0 +1,244 @@
+            try:
+                dojo_repo = data.get("dojo_repo")
+                GIT_SSH_REGEX = "^git@github.com:[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$"
+                GIT_HTTPS_REGEX = "^https://github.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$"
+                assert re.match(GIT_SSH_REGEX, dojo_repo) or re.match(GIT_HTTPS_REGEX, dojo_repo), (
+                    f"Repository violates regular expression. Must match <code>{GIT_SSH_REGEX}</code> or <code>{GIT_HTTPS_REGEX}</code>."
+                )
+
+                # clone it!
+                clone_dir = pathlib.Path(tmp_dir)/"clone"
+                clone_dir.mkdir()
+                returncode, output = sandboxed_git_clone(dojo_repo, str(ctfd_to_host_path(clone_dir)))
+
+                N=b"\n"
+                assert returncode == 0, (
+                    f"Dojo clone failed with error code {returncode}:<br>"
+                    f"<code>{output.replace(N,b'<br>').decode('latin1')}</code><br>"
+                    "Please make sure that you properly added the deploy key to the repository settings, "
+                    "and properly entered the repository URL."
+                )
+
+                # figure out the dojo ID
+                dojo_specs = list(clone_dir.glob("*.yml"))
+                assert len(dojo_specs) == 1, (
+                    f"Dojo repository must have exactly one top-level dojo spec yml named {{YOUR_DOJO_ID}}.yml. Yours has: {dojo_specs}"
+                )
+                dojo_id = dojo_specs[0].stem
+                assert id_regex(dojo_id), (
+                    f"Your dojo ID (the extensionless name of your .yml file) must be a valid URL component."
+                )
+
+                # make sure there aren't any symlink shenanigans
+                assert dojo_specs[0].is_file() and not dojo_specs[0].is_symlink(), (
+                    f"{dojo_specs[0].name} is not a regular file!"
+                )
+
+                # make sure we're not overwriting unintentionally
+                dojo_permanent_dir = pathlib.Path(DOJOS_DIR)/str(user.id)/dojo_id
+                if data.get("dojo_replace", False) not in ("true", True, 1):
+                    assert not dojo_permanent_dir.exists(), (
+                        f"You already have a cloned dojo repository containing a dojo with ID {dojo_id}."
+                    )
+
+                # make sure we're not overwriting someone else's dojo
+                existing_dojo = Dojos.query.filter_by(id=dojo_id).first()
+                if existing_dojo is not None:
+                    assert existing_dojo.owner_id == user.id, (
+                        f"A dojo with the ID {dojo_id} was already created by a different user. Please choose a different ID."
+                    )
+                    join_code = existing_dojo.join_code
+                else:
+                    join_code = random_dojo_join_code()
+
```

Comment:
Maybe we really should just not let people have the potential for name collisions on dojos.

Especially for private dojos, it's confusing to not be able to see a private dojo, and then have to just guess names until you find an available one.

We make urls for non-official dojos be something like:
`/<chosen_name>-<owner_user_id>/`

Being an official dojo can mean having a NULL owner id, getting a "prettier" url, but still keeping admin/whatever rights on your dojo.

Reply:
I'm not against that route (or `pwn.college/dojos/UID/DOJO_ID`). We could have a completely different, admin-maintained functionality to promote a dojo to the global namespace (`pwn.college/dojos/GLOBAL_DOJO_ID`) via another unique field in the dojos table.